cosa/20200805-meta-merge.md

```diff
@@ -0,0 +1,103 @@
+
+## `cosalib.meta` library
+
+`cmd-meta` currently uses the [`cosalib.meta`](https://github.com/coreos/coreos-assembler/blob/master/src/cosalib/meta.py) library for reading and writing `meta.json`. To support merging of `meta.json` files, the library will be updated to allow for merging in-memory updates with on-disk content.
+
+Only `meta.json` that is validated will be merged under the following rules:
+- the on-disk version must match in core details for name and buildid for in memory model of the meta
+- on-disk meta.json takes precedence over the in-memory meta
+- all versions must be validated before and after merging before commiting to disk.
+
+Prior to merging `cosalib.meta` will lock the localfile.
+
+## build mode denoted in `meta.json`
+
+Two new build "modes" will be added. `cmd-build` will have build modes written in the `meta.json` of `parallel` (default) or distributed.
+
+```
+{ 
+    "cosa.buildmode": "<parallel|distributed>"
+}
+```
+
+`cmd-build` will have the flag of `--distributed` added to set the build mode.
+
+## Distributed builds
+
+When `cosa.buildmode` is set to `distributed`, `cmd-buildextend.<ARTIFACT>` commands will create suffixed `meta.json.<ARTIFACT>` instead of writing directly to `meta.json`. A new command called `cmd-meta-merge` will be introduced that will read `meta.json.<ARTIFACT>` and will ONLY combine the images and the cloud-build specific sections into `meta.json`. 
+
+Pipelines orchestrating the build will be responsible for shuffling artifacts around. 
+
+All `meta.json.<ARTIFACT>` files will be archived in S3 alongside the `meta.json`. 
+
+`cmd-meta-merge` will be able to read `meta.json.<ARTIFACT>` located directly in S3. `cmd-meta-merge` will:
+- read in `meta.json` and validate it.
+- iterate over the remaining `meta.json.<ARTIFACT>`. Validation against the schema will happen during the read and after merging.
+
+## Parallel builds
+
+When `cosa.buildmode` is set to "parallel", the `cmd-buildextend-*` commands will manage the merging of `meta.json` in real time using the `cosalib.meta` Python code to prevent racing conditions. To support bash commands `finalize-artifact` will be replaced with `cmd-meta-merge`.
+
+This mode assumes that the build happens within the same kernel namespace. 
+
+## Validation Requirement
+
+All merged `meta.json` must be validated against the schema.
+
+## Publication
+
+Orchestrating pipelines are responsible for only publishing builds to the canonical locations.
+
+## Optional `meta.json`
+
+The schema will be updated to allow for `publisher` meta-data, i.e.
+```
+{ 
+    "publisher": {
+        ...
+    }
+}
+```
+
+This meta-data section is indended for use by publishers. COSA will:
+- ignore the section entirely
+- the schema will have it added as an optional, albeit ignored field.
+
+And example might be:
+```
+{
+    "publisher": {
+        "state": "done",
+        "developer-build": False
+    }
+}
+```
```

Comment:
Hmm, do you have more context on this one? It sounds like pipeline implementation details leaking into the schema. Can we drop this part of the proposal until it's actually needed?

Reply:
dropped